src/cockpit/channels/dbus.py

```diff
@@ -172,11 +171,47 @@ class DBusChannel(Channel):
     matches = None
     name = None
     bus = None
+    owner = None
 
     # This needs to be a fair mutex so that outgoing messages don't
     # get re-ordered.  asyncio.Lock is fair.
     watch_processing_lock = asyncio.Lock()
 
+    async def setup_name_owner_tracking(self):
+        def send_owner(owner):
+            # We must be careful not to send duplicate owner
+            # notifications. cockpit.js relies on that.
+            if self.owner != owner:
+                self.owner = owner
+                self.send_message(owner=owner)
+
+        def handler(message):
+            name, old, new = message.get_body()
+            send_owner(owner=new if new != "" else None)
+        rule = f"type='signal',sender='org.freedesktop.DBus',path='/org/freedesktop/DBus',interface='org.freedesktop.DBus',member='NameOwnerChanged',arg0='{self.name}'"
```

Comment:
You seem to be putting `rule` on its own line to avoid excessive line length, but I don't think it's helping much.

I think it might be nice to do something like:

```python
   rule = {
       'type': 'signal',
       'sender': 'org.freedesktop.DBus',
       ...
    }
    rule_text = ','.join(f"{key}='{val}'" for key, val in rule.items())
    self.matches.append(self.bus.add_match(rule_text, handler))
```

or so, and even to have a proper internal API for that.  IIRC, we need to do that kind of thing in response to match rules from the user, already.

You could even add a `**kwargs` on your internal API and then call it like

```python
    self.my_magic_matchy_thing(type='signal', sender='org.freedesktop.DBus', ...)
```

Reply:
This is done better in #18076.